File: Editorial/132_Palindrome_Partitioning_II/CSharp/2_Dynamic_Programming_Top_Down.cs
public class Solution {
    private int?[][] memoCuts;
    private bool?[][] memoPalindrome;

    public int MinCut(string s) {
        memoCuts = new int ?[s.Length][];
        memoPalindrome = new bool ?[s.Length][];
        for (int i = 0; i < s.Length; i++) {
            memoCuts[i] = new int?[s.Length];
            memoPalindrome[i] = new bool?[s.Length];
        }

        return FindMinimumCut(s, 0, s.Length - 1, s.Length - 1).Value;
    }

    private int? FindMinimumCut(string s, int start, int end, int minimumCut) {
        // base case
        if (start == end || IsPalindrome(s, start, end).Value) {
            return 0;
        }

        // check for results in memoCuts
        if (memoCuts[start][end].HasValue) {
            return memoCuts[start][end].Value;
        }

        for (int currentEndIndex = start; currentEndIndex <= end;
             currentEndIndex++) {
            if (IsPalindrome(s, start, currentEndIndex).Value) {
                minimumCut = Math.Min(
                    minimumCut,
                    1 + FindMinimumCut(s, currentEndIndex + 1, end, minimumCut)
                            .Value);
            }
        }

        return memoCuts[start][end] = minimumCut;
    }

    private bool? IsPalindrome(string s, int start, int end) {
        if (start >= end) {
            return true;
        }

        // check for results in memoPalindrome
        if (memoPalindrome[start][end] != null) {
            return memoPalindrome[start][end].Value;
        }

        return memoPalindrome[start][end] =
                   (s[start] == s[end]) &&
                   IsPalindrome(s, start + 1, end - 1).Value;
    }
}

File: Editorial/132_Palindrome_Partitioning_II/JavaScript/2_Dynamic_Programming_Top_Down.js
var minCut = function (s) {
    const n = s.length;
    const memoCuts = Array.from({ length: n }, () => Array(n).fill(-1));
    const memoPalindrome = Array.from({ length: n }, () => Array(n).fill(null));

    // Function to check if the substring is a palindrome
    const isPalindrome = (start, end) => {
        if (start >= end) return true;
        if (memoPalindrome[start][end] !== null)
            return memoPalindrome[start][end];
        memoPalindrome[start][end] =
            s[start] === s[end] && isPalindrome(start + 1, end - 1);
        return memoPalindrome[start][end];
    };

    // Function to find the minimum cut for palindrome partitioning
    const findMinimumCut = (start, end, minimumCut) => {
        // base case
        if (start === end || isPalindrome(start, end)) {
            return 0;
        }

        // check for results in memoCuts
        if (memoCuts[start][end] !== -1) {
            return memoCuts[start][end];
        }

        for (
            let currentEndIndex = start;
            currentEndIndex <= end;
            currentEndIndex++
        ) {
            if (isPalindrome(start, currentEndIndex)) {
                minimumCut = Math.min(
                    minimumCut,
                    1 + findMinimumCut(currentEndIndex + 1, end, minimumCut),
                );
            }
        }
        memoCuts[start][end] = minimumCut;
        return minimumCut;
    };

    return findMinimumCut(0, n - 1, n - 1);
};
